tests/subsys/net/lib/nrf_cloud/src/main.c

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <unity.h>
+#include <zephyr.h>
+#include <stdio.h>
+#include <net/nrf_cloud.h>
+#include <mock_nrf_cloud_transport.h>
+#include "nrf_cloud_fsm.h"
+
+/* This is required since unity_main return int and
+ * zephyr expects main not to return any value
+ */
+extern int unity_main(void);
+
+/* The library can only be initialized when the internal state is
+ * idle, so the test verifies that nrf_cloud_init function will
+ * return with permission denied error (-EACCES) when not idle
+ */
+void test_init_not_idle(void)
+{
+	struct nrf_cloud_init_param params = {
+		.event_handler = NULL,
+		.client_id = NULL,
+	};
+
+	/* nfsm_set_current_state_and_notify directly sets the
+	 * internal state of the nrf_cloud lib, allowing us to
+	 * easily test various initialization scenarios
+	 */
+	nfsm_set_current_state_and_notify(STATE_TOTAL, NULL);
+	TEST_ASSERT_EQUAL_MESSAGE(-EACCES, nrf_cloud_init(&params),
+		"return should be -EACCES when not idle");
+}
+
+/* Verify that nrf_cloud_init function will throw an error
+ * of invalid argument (-EINVAL) when given null params
+ */
+void test_init_null_params(void)
+{
+	/* Set the internal state to idle */
+	nfsm_set_current_state_and_notify(STATE_IDLE, NULL);
```

Comment:
Rather than set `STATE_IDLE` directly, I suggest calling `nrf_cloud_uninit` after every test, ensuring it will always be in an un-initialized state.

Then, in this test, _assert_ that `nfsm_get_current_state` returns IDLE

This way you minimize directly interfering with the state machine

I suggest you do this in a tearDown function, which I **_think_** you can create by literally just having a function named `void tearDown(void)` -- better confirm that such a function actually runs if you declare it.

You can do the same for other tests